### PR TITLE
Bulk Load CDK: Preliminary Refactor before adding mappers

### PR DESCRIPTION
**Pull request by johnny-schmidt**
This is in preparation for a change that adds the schema/value mappers in `io.airbyte.cdk.load.data` to the object storage formatted writers for avro and parquet. Combined, these make the next change less verbose / easier to test.

### What
* Remove `NullType` from `AirbyteType`
### Why
* It creates ambiguity (`FieldType(StringType, nullable=true)` == `UnionType(StringType, NullType)`; there should be exactly one way to represent everything
* Especially: this ambiguity makes it difficult to reason about mapper chains for avro/parquet mangling; it forces the converters to keep inspecting the object to see what type of nullable we're dealing with
* It also requires an additional mapper pass `NullableToUnionNull` to resolve the ambiguity (now we can just drop it)
* `NullType` at the field level isn't legal, even tho we see it sometimes; `AirbyteType` should only contain what is legal

### What
* Move conversions to dedicated classes with extension function interfaces
* Specifically change the RecordDecorator so that it can be used with an extension function
### Why
* I'm following specific design principles here:
  *  separate use cases (conversions) from data classes; otherwise the data classes pile up with nested static functions implementing multiple concerns (we don't follow this consistently yet, but IMO we should)
  * conversions are all classes with standard naming (`{Source}To{Dest}`) so they're easily searchable
  * extension functions allow for a more fluent style in the client code

### CODE REVIEW

File: airbyte-cdk/bulk/core/load/src/test/kotlin/io/airbyte/cdk/load/data/json/AirbyteSchemaTypeToJsonSchemaTest.kt [hunk 1]
```diff
@@ -4,145 +4,172 @@
...
-
-        val timeTypeFieldNames =
-            listOf("time", "timestamp", "time_without_timezone", "timestamp_without_timezone")
-        timeTypeFieldNames.forEach { fieldName ->
-            val expected = props.get(fieldName) as ObjectNode
-            if (listOf("date-time", "time").contains(expected.get("format").asText())) {
-                val formatName = expected.get("format").asText().replace("date-time", "timestamp")
-                if (!expected.has("airbyte_type")) {
-                    expected.put("airbyte_type", "${formatName}_with_timezone")
+    // A json of every supported type
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/test/kotlin/io/airbyte/cdk/load/data/json/AirbyteSchemaTypeToJsonSchemaTest.kt:26]: it was easier to rewrite this than to try to remediate it

(this style is extremely copilot friendly, so it was close to a 1-shot)

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteSchemaIdentityMapper.kt [hunk 2]
```diff
@@ -4,10 +4,13 @@
 
 package io.airbyte.cdk.load.data
 
-interface AirbyteSchemaIdentityMapper {
-    fun map(schema: AirbyteType): AirbyteType =
+interface AirbyteSchemaMapper {
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteSchemaIdentityMapper.kt:7]: This leaked in from the next change. But the point is to make the mappers generalizable for chaining

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/check/DestinationChecker.kt [hunk 3]
```diff
@@ -25,7 +25,7 @@ interface DestinationChecker<C : DestinationConfiguration> {
         DestinationStream(
             descriptor = DestinationStream.Descriptor("testing", "test"),
             importType = Append,
-            schema = NullType,
+            schema = ObjectTypeWithoutSchema,
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/check/DestinationChecker.kt:28]: tangential, but I suspect we'll eventually change this to something like `ObjectType(mapOf("foo" to StringType))` (mostly for warehouse destinations, where creating a table with no columns is probably dumb)

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/UnionTypeToDisjointRecord.kt [hunk 4]
```diff
@@ -8,30 +8,24 @@ import io.airbyte.cdk.load.message.DestinationRecord
 
 class UnionTypeToDisjointRecord : AirbyteSchemaIdentityMapper {
     override fun mapUnion(schema: UnionType): AirbyteType {
-        val (nullOptions, nonNullOptions) = schema.options.partition { it is NullType }
-        if (nonNullOptions.size < 2) {
+        if (schema.options.size < 2) {
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/UnionTypeToDisjointRecord.kt:11]: unrelated thought:
* if size == 1 -> should we just return `schema.options.first()`?
* if size == 0 -> wtf :P (uh, maybe return UnknownType? `{oneOf: []}` is a pretty stupid schema)

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonSchemaToAirbyteType.kt [hunk 5]
```diff
@@ -24,33 +26,33 @@ class JsonSchemaToAirbyteType {
                     "number" -> fromNumber(schema)
                     "array" -> fromArray(schema)
                     "object" -> fromObject(schema)
-                    "null" -> NullType
+                    "null" -> null
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonSchemaToAirbyteType.kt:29]: should we treate `{type: null}` schemas as just UnknownType?

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteSchemaIdentityMapper.kt (hunk 2 above)
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteSchemaIdentityMapper.kt:7]: is the idea that we only strictly _need_ the `map` method for chaining, and all the mapXyz methods are just for convenience?

**Review by edgao — APPROVED** [review]
lgtm, had some unrelated thoughts - nothing blocking

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteSchemaIdentityMapper.kt (hunk 2 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteSchemaIdentityMapper.kt:7]: yeah, I think `IdentityMapper` is the wrong name, it's more like TypeAwareMapper or something. So `Mapper => just `map`. The IdentityMapper class has all the bells and whistles.

In practical terms, I needed a common parent type for the Noop and visiting value mappers, so I broke out an interface for both of them for symmetry

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/UnionTypeToDisjointRecord.kt (hunk 4 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/UnionTypeToDisjointRecord.kt:11]: In theory nothing should get to this point with a singleton Union

actually singleton unions shouldn't be representable at all, ideally

like ideally the type classes would be set up in a way where you're always getting an `AirbyteType` when you try to create a `Union`, and
* singletons just become the type
* redundancies disappear into `Set<Options>`

Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/UnionTypeToDisjointRecord.kt:11]: Which means that this guard is probably a little vestigial. :) It only made sense when there were null types.

I'll remove it in a cleanup pass.

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonSchemaToAirbyteType.kt (hunk 5 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonSchemaToAirbyteType.kt:29]: That makes sense at the field level, but it's still legal(ish?) Airbyte JsonSchema to represent a nullable as `"type": ["null", "string"]`

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonSchemaToAirbyteType.kt [hunk 6]
```diff
@@ -107,60 +110,48 @@ class JsonSchemaToAirbyteType {
...
                 } else {
-                    convert(it)
+                    convertInner(it)
                 }
             }
-        return UnionType(unionOptions)
+        return optionsToUnionOrSingleType(unionOptions)
     }
+
+    private fun optionsToUnionOrSingleType(options: List<AirbyteType>): AirbyteType =
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonSchemaToAirbyteType.kt:149]: Pursuant to early comment, basically move this into a companion object in the union type, and block direct construction

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonSchemaToAirbyteType.kt (hunk 5 above)
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonSchemaToAirbyteType.kt:29]: (not blocking, mostly just whinging about jsonschema)

don't we treat everything as nullable anyway? i.e. from our pov, `{type: string} == {type: [null, string]}`

I almost want to have some kind of preprocess somewhere, where we convert all type decls to the array form (`{type: foo}` -> `{type: [foo]}`) and then filter out the `"null"` elements from those arrays

(and then `{type: []}` becomes UnknownType, and we don't need to handle null anywhere)

**Comment by johnny-schmidt** [comment]
@tryangul @benmoriceau I'm going to merge this in, since it's been open for a day. I'll keep adding you all, jump in as you have bandwidth

**Pull request merged**